README.md
# RFISO8601DateTime

[![CI Status](http://img.shields.io/travis/readefries/RFISO8601DateTime.svg?style=flat)](https://travis-ci.org/readefries/RFISO8601DateTime)
[![Version](https://img.shields.io/cocoapods/v/RFISO8601DateTime.svg?style=flat)](http://cocoapods.org/pods/RFISO8601DateTime)
[![License](https://img.shields.io/cocoapods/l/RFISO8601DateTime.svg?style=flat)](http://cocoapods.org/pods/RFISO8601DateTime)
[![Platform](https://img.shields.io/cocoapods/p/RFISO8601DateTime.svg?style=flat)](http://cocoapods.org/pods/RFISO8601DateTime)

## Usage

Using is very easy, to parse an RFC3339 date/time string, just do the following.

```
let rfc3339DateTime = "1985-04-12T23:20:50.52Z"
let parsedDateTime = NSDate.parseDateString(rfc3339DateTime)
```

## Supported date/time formats

RFISO8601DateTime currently supports the following date/time formats:
* calendar month (e.g., 2008-08).
* Calendar date, such as 2008-08-30 or 20080830.
* Ordinal date (e.g., 2008-243). 
* Week of the year (e.g., 2008-W35).
* Week date (e.g., 2008-W35-6).
* Hours and minutes (e.g., 17:21).
* Hours, minutes, and seconds (e.g., 17:21:59).
* Time zone designator (e.g., Z, +07 or +07:00).
* Hours, minutes, and seconds with time zone designator (e.g., 17:21:59+07:00). 
* Calendar date with hours, minutes, and seconds (e.g., 2008-08-30 17:21:59 or 20080830 172159).
* Date, with optional time zone (e.g., 2008-08-30 or 2008-08-30+07:00).
* Time, with optional fractional seconds and time zone (e.g., 01:45:36 or 01:45:36.123+07:00).
* Date and time, with optional fractional seconds and time zone (e.g., 2008-08-30T01:45:36 or 2008-08-30T01:45:36.123Z).
* Date and time in RFC2822 format (e.g. Fri, 21 Nov 1997 09:55:06 -0600).

## Requirements

## Installation

RFISO8601DateTime is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```
pod 'RFISO8601DateTime', '~> 3.0' # Swift 4.0
```

```
pod 'RFISO8601DateTime', '~> 2.0' # Swift 3.0
```

```
pod 'RFISO8601DateTime', '~> 1.0' # Swift 2.3
```

## Author

Hindrik Bruinsma, de@readefries.nl

## License

RFISO8601DateTime is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# 3.0.3

* Updated sources for Xcode 10 and Swift 5

# 3.0.0

* Updated sources for Xcode 9 and Swift 4

# 2.0.2

* Removed a lot of force unwraps, and replaced them with if let/guard lets to prevent possible crashes

# 2.0.1

* Fixed issue which would cause crashes for certain date formats

# 2.0.0

* Upgrade to Swift 3.0

# 1.0.0

* Upgraded to Swift 2.3

# 0.2.2

* Added support for RFC2822 date/time notation 

# 0.1.1

* Added travis support

# 0.1.0

* Inial Swift rebuild from my previous Objective-C verion.
* Added travis support
* Added library to Cocoapods


RFISO8601DateTime/NSDate+ISO8601.swift
import Foundation

public extension Date {
  
  static func parseDateString(_ dateTimeString: String) -> Date? {

    var date: Date? = nil
    var time: Date? = nil
    var timeZone: TimeZone? = nil
    
    // Date and time

    // Date and time, with fractional seconds and time zone
    if let _ = dateTimeString.range(of: ISO8601Constants.DateTimeWithFractionalSecondsAndTimeZoneRegexp, options: .regularExpression) {
      if let dateRange = dateTimeString.range(of: ISO8601Constants.DateWithTimeZoneRegexp, options: .regularExpression) {
        date = dateWithString("\(dateTimeString[dateRange])")
      }

      if let timeRange = dateTimeString.range(of: ISO8601Constants.TimeWithFractionalSecondsRegexp, options: .regularExpression) {
        time = timeWithString("\(dateTimeString[timeRange])")
      }

      if let timezoneRange = dateTimeString.range(of: ISO8601Constants.TimeZoneDesignatorRegexp, options: .regularExpression) {
        timeZone = TimeZone.timeZoneWithString("\(dateTimeString[timezoneRange])")
      }

      return combineDateTimeTimezone(date, time: time, timezone: timeZone)
    }
    
    // Calendar date with hours, minutes, and seconds (e.g., 2008-08-30 17:21:59 or 20080830 172159).
    if let _ = dateTimeString.range(of: ISO8601Constants.CalenderDateHoursMinutesSecondsRegexp, options: .regularExpression) {
      if let dateRange = dateTimeString.range(of: ISO8601Constants.CalendarDateRegexp, options: .regularExpression) {
        date = dateWithString("\(dateTimeString[dateRange])")
      }
      
      if let timeRange = dateTimeString.range(of: ISO8601Constants.HoursMinutesSecondsRegexp, options: .regularExpression) {
        time = timeWithString("\(dateTimeString[timeRange])")
      }

      guard let date = date, let time = time else { return nil }

      return combineDateTime(date, time: time)
    }

    // RFC2822 Date
    if let _ = dateTimeString.range(of: ISO8601Constants.DateTimeRFC2822Regexp, options: .regularExpression) {
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.DateTimeRFC2822Format
      
      return ISO8601Constants.dateFormatter.date(from: dateTimeString)
    }
    
    if let range = dateTimeString.range(of: ISO8601Constants.WeekDateRegexp, options: .regularExpression) {
      return dateWeekWithSTring("\(dateTimeString[range])")
    }
      // Ordinal date e.g. 2016-021
    else  if let _ = dateTimeString.range(of: ISO8601Constants.OrdinalDateRegexp, options: .regularExpression) {
      let dateTimeStringWithoutDashes = dateTimeString.replacingOccurrences(of: "-", with: "")
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.OrdinalDateFormat
      
      return ISO8601Constants.dateFormatter.date(from: dateTimeStringWithoutDashes)
    }
    
    if let dateRange = dateTimeString.range(of: ISO8601Constants.DateWithTimeZoneRegexp, options: .regularExpression) {
      date = dateWithString("\(dateTimeString[dateRange])")
    }
    
    if let timeRange = dateTimeString.range(of: ISO8601Constants.TimeWithFractionalSecondsRegexp, options: .regularExpression) {
      time = timeWithString("\(dateTimeString[timeRange])")
    }
    
    if let timezoneRange = dateTimeString.range(of: ISO8601Constants.TimeZoneDesignatorRegexp, options: .regularExpression) {
      timeZone = TimeZone.timeZoneWithString("\(dateTimeString[timezoneRange])")
    }
    
    // Only try to parse week, if now date and time has been found
    if date == nil && time == nil, let timeZone = timeZone {
      var calendar = Calendar.current
      calendar.timeZone = timeZone

      let timeComponents = DateComponents()
      (timeComponents as NSDateComponents).timeZone = timeZone
      
      return calendar.date(from: timeComponents)
    }
    else if let date = date, time == nil {
      if let timeZone = timeZone {
        var calendar = Calendar.current
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        
        let dateComponents = (calendar as NSCalendar).components([.year, .month, .day], from: date)
        (dateComponents as NSDateComponents).timeZone = timeZone
        
        return calendar.date(from: dateComponents)
      }
      else {
        return date
      }
    }
    else if let time = time, date == nil {
      if let timeZone = timeZone {
        return combineDateTimeTimezone(date, time: time, timezone: timeZone)
      }
      else {
        return time
      }
    }
    
    return nil
  }
  
  fileprivate static func combineDateTime(_ date: Date, time: Date) -> Date {
    return combineDateTimeTimezone(date, time: time, timezone: nil)
  }
  
  fileprivate static func combineDateTimeTimezone(_ date: Date?, time: Date?, timezone: TimeZone?) -> Date {
    
    var calendar = Calendar.current
    calendar.timeZone = TimeZone(secondsFromGMT: 0)!

    var mergedComponents = DateComponents()

    if let date = date {
      let dateComponents = (calendar as NSCalendar).components([.year, .month, .day], from: date)

      mergedComponents.year = dateComponents.year
      mergedComponents.month = dateComponents.month
      mergedComponents.day = dateComponents.day
    }

    if let time = time {
      let timeComponents = (calendar as NSCalendar).components([.hour, .minute, .second, .nanosecond], from: time)

      mergedComponents.hour = timeComponents.hour
      mergedComponents.minute = timeComponents.minute
      mergedComponents.second = timeComponents.second
      mergedComponents.nanosecond = timeComponents.nanosecond

    }

    if let timezone = timezone {
      mergedComponents.timeZone = timezone
    }

    return calendar.date(from: mergedComponents)!
  }

  fileprivate static func dateWeekWithSTring(_ dateWeekString: String) -> Date? {

    // Week of the year e.g. 2016-W03
    if let _ = dateWeekString.range(of: ISO8601Constants.WeekOfTheYearRegexp, options: .regularExpression) {
      let dateWeekStringWithoutDashes = dateWeekString.replacingOccurrences(of: "-", with: "")
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.WeekOfTheYearFormat
      
      return ISO8601Constants.dateFormatter.date(from: dateWeekStringWithoutDashes)
    }
    
    // Week date e.g. 2016-W03-6
    if let _ = dateWeekString.range(of: ISO8601Constants.WeekDateRegexp, options: .regularExpression) {
      let dateWeekStringWithoutDashes = dateWeekString.replacingOccurrences(of: "-", with: "")
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.WeekDateFormat
      
      return ISO8601Constants.dateFormatter.date(from: dateWeekStringWithoutDashes)
    }
    
    return nil
  }
  
  fileprivate static func dateWithString(_ dateString: String) -> Date? {
    
    // Calender month e.g. 2016-01
    if let _ = dateString.range(of: ISO8601Constants.CalendarMonthRegexp, options: .regularExpression) {
      let dateStringWithoutDashes = dateString.replacingOccurrences(of: "-", with: "")
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.CalendarMonthFormat
      
      return ISO8601Constants.dateFormatter.date(from: dateStringWithoutDashes)
    }
    
    // Calender date e.g. 2016-01-21
    if let _ = dateString.range(of: ISO8601Constants.CalendarDateRegexp, options: .regularExpression) {
      let dateStringWithoutDashes = dateString.replacingOccurrences(of: "-", with: "")
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.CalendarDateFormat
      
      return ISO8601Constants.dateFormatter.date(from: dateStringWithoutDashes)
    }
    
    return nil
  }
  
  fileprivate static func timeWithString(_ timeString: String) -> Date? {

    // Hours and minutes e.g. 10:10:34
    if let _ = timeString.range(of: ISO8601Constants.HoursMinutesSecondsRegexp, options: .regularExpression) {
      let timeStringWithoutDashes = timeString.replacingOccurrences(of: ":", with: "")
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.HoursMinutesSecondsFormat

      return ISO8601Constants.dateFormatter.date(from: timeStringWithoutDashes)
    }

    // Hours and minutes e.g. 10:10
    if let _ = timeString.range(of: ISO8601Constants.HoursMinutesRegexp, options: .regularExpression) {
      let timeStringWithoutDashes = timeString.replacingOccurrences(of: ":", with: "")
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.HoursMinutesFormat
      
      return ISO8601Constants.dateFormatter.date(from: timeStringWithoutDashes)
    }

    // Time, with fractional seconds (e.g. 01:45:36.123).
    if let _ = timeString.range(of: ISO8601Constants.TimeWithFractionalSecondsRegexp, options: .regularExpression) {
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.TimeWithFractionalSecondsFormat
      
      return ISO8601Constants.dateFormatter.date(from: timeString)
    }
    
    // Time, with fractional seconds and time zone (e.g. 01:45:36.123+07:00).
    if let _ = timeString.range(of: ISO8601Constants.TimeWithFractionalSecondsAndTimeZoneRegexp, options: .regularExpression) {
      ISO8601Constants.dateFormatter.dateFormat = ISO8601Constants.TimeWithFractionalSecondsAndTimeZoneFormat
      
      return ISO8601Constants.dateFormatter.date(from: timeString)
    }
    
    return nil
  }
}


RFISO8601DateTime/RFISO8601DateTime.swift

import Foundation

struct ISO8601Constants {
  
  static let CalendarMonthRegexp = "^([0-9]{4})-(1[0-2]|0[1-9])$"
  static let CalendarMonthFormat = "yyyyMM"
  
  static let CalendarDateRegexp = "([0-9]{4})-?(1[0-2]|0[1-9])-?(3[01]|0[1-9]|[12][0-9])"
  static let CalendarDateFormat = "yyyyMMdd"
  
  static let OrdinalDateRegexp = "([0-9]{4})-?(36[0-6]|3[0-5][0-9]|[12][0-9]{2}|0[1-9][0-9]|00[1-9])"
  static let OrdinalDateFormat = "yyyyddd"
  
  static let WeekOfTheYearRegexp = "^([0-9]{4})-?W(5[0-3]|[1-4][0-9]|0[1-9])$"
  static let WeekOfTheYearFormat = "yyyy'W'ww"
  
  static let WeekDateRegexp = "([0-9]{4})-?W(5[0-3]|[1-4][0-9]|0[1-9])?-?([1-7])?"
  static let WeekDateFormat = "yyyy'W'wwc"
  
  static let HoursMinutesRegexp = "(2[0-3]|[01][0-9]):?([0-5][0-9])$"
  static let HoursMinutesFormat = "HHmm"
  
  static let HoursMinutesSecondsRegexp = "(2[0-3]|[01][0-9]):?([0-5][0-9]):([0-5][0-9])$"
  static let HoursMinutesSecondsFormat = "HHmmss"
  
  static let TimeZoneDesignatorRegexp = "(Z|[+-](?:2[0-3]|[01][0-9])(?::?(?:[0-5][0-9]))?)$"
  static let TimeZoneDesignatorFormat = "X"
  
  static let HoursMinutesSecondsWithTimeZoneDesignatorRegexp = "(2[0-3]|[01][0-9]):?([0-5][0-9]):?([0-5][0-9])(Z|[+-](?:2[0-3]|[01][0-9])(?::?(?:[0-5][0-9]))?)"
  static let HoursMinutesSecondsWithTimeZoneDesignatorFormat = "hhmmssX"
  
  static let CalenderDateHoursMinutesSecondsRegexp = "([0-9]{4})-?(1[0-2]|0[1-9])-?(3[01]|0[1-9]|[12][0-9])( |T)(2[0-3]|[01][0-9]):?([0-5][0-9]):?([0-5][0-9])"
  static let CalenderDateHoursMinutesSecondsFormat = "yyyyMMdd' 'hhmmss"
  
  static let DateWithTimeZoneRegexp = "(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])(-)?(3[01]|0[1-9]|[12][0-9])?(Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9])?"
  static let DateWithTimeZoneFormat = "yyyyMMddx"
  
  static let TimeWithFractionalSecondsAndTimeZoneRegexp = "(2[0-3]|[01][0-9]):([0-5][0-9])(:)?([0-5][0-9])?(\\.[0-9]+)?(Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9])?"
  static let TimeWithFractionalSecondsAndTimeZoneFormat = "HH:mm:ss'.'SSSZZZZ"
  
  static let TimeWithFractionalSecondsRegexp = "(2[0-3]|[01][0-9]):([0-5][0-9]):?([0-5][0-9])?(\\.[0-9]+)?"
  static let TimeWithFractionalSecondsFormat = "HH:mm:ss'.'SSS"
  
  static let DateTimeWithFractionalSecondsAndTimeZoneRegexp = "(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])( |T)(2[0-3]|[01][0-9]):([0-5][0-9]):([0-5][0-9])(\\.[0-9]+)?(Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9])"
  static let DateTimeWithFractionalSecondsAndTimeZoneFormat = "yyyyMMdd'T'HHmmss'.'SSSX"
  
  static let DateTimeRFC2822Regexp = "^(?:(Sun|Mon|Tue|Wed|Thu|Fri|Sat),\\s+)?(0[1-9]|[1-2]?[0-9]|3[01])\\s+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\\s+(19[0-9]{2}|[2-9][0-9]{3})\\s+(2[0-3]|[0-1][0-9]):([0-5][0-9])(?::(60|[0-5][0-9]))?\\s+([-\\+][0-9]{2}[0-5][0-9]|(?:UT|GMT|(?:E|C|M|P)(?:ST|DT)|[A-IK-Z]))(\\s+|\\(([^\\(\\)]+|\\\\\\(|\\\\\\))*\\))*$"
  static let DateTimeRFC2822Format = "EEE', 'dd' 'MMM' 'yyyy' 'HH':'mm':'ss' 'Z"
  
  static let dateFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    
    return formatter
  }()
  
  static var numberFormatter: NumberFormatter = {
    let formatter = NumberFormatter()
    
    return formatter
  }()
}


RFISO8601DateTime/NSTimeZone+ISO8601.swift

import Foundation

public extension TimeZone {
  // Time zone designator e.g., Z, +01 or +01:00
  static func timeZoneWithString(_ dateString: String) -> TimeZone! {
    
    var timeZoneString: String!
    
    if let rangeOfTimezone = dateString.range(of: ISO8601Constants.TimeZoneDesignatorRegexp, options: .regularExpression) {
      timeZoneString = String(dateString[rangeOfTimezone.lowerBound..<rangeOfTimezone.upperBound])

      if timeZoneString == "Z" {
        return TimeZone(secondsFromGMT: 0)
      }
      
      if let _ = timeZoneString.range(of: ISO8601Constants.TimeZoneDesignatorRegexp, options: .regularExpression) {
        timeZoneString = timeZoneString.replacingOccurrences(of: ":", with: "")

        let hoursString = "\(timeZoneString[timeZoneString.index(timeZoneString.startIndex, offsetBy: 1)..<timeZoneString.index(timeZoneString.startIndex, offsetBy: 3)])"
        var minutesString: String! = nil
        
        if timeZoneString.count == 5 {
          minutesString = "\(timeZoneString[timeZoneString.index(timeZoneString.startIndex, offsetBy: 3)..<timeZoneString.index(timeZoneString.startIndex, offsetBy: 5)])"
        }
        
        let formatter = NumberFormatter()
        
        let hours = formatter.number(from: hoursString)
        let minutes = minutesString == nil ? (0) : formatter.number(from: minutesString as String)
        
        var seconds = ((hours?.intValue)! * 3600) + ((minutes?.intValue)! * 60)

        if timeZoneString.firstIndex(of: "\u{2d}") != nil {
          seconds = seconds * -1
        }
        
        return TimeZone(secondsFromGMT: seconds)
      }
    }
    
    return nil
  }
}


RFISO8601DateTimeTests/RFISO8601DateTimeTests.swift

import XCTest

import RFISO8601DateTime

class ISO8601DateTimeTests: XCTestCase {
  
  override func setUp() {
    super.setUp()
    // Put setup code here. This method is called before the invocation of each test method in the class.
  }
  
  override func tearDown() {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    super.tearDown()
  }
  
  func testInvalidDate() {
    let invalidDate = "invalidDate"
    
    XCTAssertNil(Date.parseDateString(invalidDate))
  }
  
  func testCalendarMonth() {
    let calendarMonth = "2016-01"

    if let parsedDate = Date.parseDateString(calendarMonth) {
      let calendar = Calendar.current
      let components = (calendar as NSCalendar).components([.month , .year], from: parsedDate)
      
      XCTAssertTrue(components.year == 2016, String(format: "The year should be 2016, not %d", components.year!))
      XCTAssertTrue(components.month == 1, String(format: "The month should be 1, not %d", components.month!))
    }
  }
  
  func testCalendarDate() {
    let calendarMonth = "2016-01-21"

    if let parsedDate = Date.parseDateString(calendarMonth) {
      let calendar = Calendar.current
      let components = (calendar as NSCalendar).components([.day , .month , .year], from: parsedDate)

        XCTAssertTrue(components.year == 2016, String(format: "The year should be 2016, not %d", components.year!
        ))
        XCTAssertTrue(components.month == 1, String(format: "The month should be 1, not %d", components.month!))
        XCTAssertTrue(components.day == 21, String(format: "The day should be 21, not %d", components.day!))
    }
  }
  
  func testOrdinalDate() {
    let ordinalDate = "2016-021"

    if let parsedDate = Date.parseDateString(ordinalDate) {
      let calendar = Calendar.current
      let components = (calendar as NSCalendar).components([.year], from: parsedDate)
      let day = (calendar as NSCalendar).ordinality(of: .day, in: .year, for: parsedDate)
      
      XCTAssertTrue(components.year == 2016, String(format: "The year should be 2016, not %d", components.year!))
      XCTAssertTrue(day == 21, String(format: "The day should be 21, not %d", day))
    }
  }
  
  func testWeekDate() {
    let weekDate = "2016-W03-6"

    if let parsedDate = Date.parseDateString(weekDate) {
      let calendar = Calendar.current
      let components = (calendar as NSCalendar).components([.weekOfYear, .weekday, .year], from: parsedDate)
      
      XCTAssertTrue(components.year == 2016, String(format: "The year should be 2016, not %d", components.year!))
      XCTAssertTrue(components.weekOfYear == 3, String(format: "The week of the year should be 3, not %d", components.weekOfYear!))
      XCTAssertTrue(components.weekday == 6, String(format: "The day of the week should be 6, not %d", components.weekday!))
    }
  }
  
  func testWeekOfTheYear() {
    let weekOfTheYear = "2016-W03"

    if let parsedDate = Date.parseDateString(weekOfTheYear) {
      let calendar = Calendar.current
      let components = (calendar as NSCalendar).components([.weekOfYear , .year], from: parsedDate)
      
      XCTAssertTrue(components.year == 2016, String(format: "The year should be 2016, not %d", components.year!))
      XCTAssertTrue(components.weekOfYear == 3, String(format: "The week of the year should be 3, not %d", components.weekOfYear!))
    }
  }
  
  func testTimeWithFractionalSecondsAndTimezone() {
    let time = "01:45:36.123+07:00"
    
    if let parsedDate = Date.parseDateString(time) {
      
      if let timeZone = TimeZone.timeZoneWithString(time) {
        
        var calendar = Calendar.current
        calendar.timeZone = timeZone
        let components = (calendar as NSCalendar).components([.hour, .minute, .second, .nanosecond, .timeZone], from: parsedDate)

        XCTAssertTrue(components.hour == 1, String(format: "The hour should be 1, not %d", components.hour!))
        XCTAssertTrue(components.minute == 45, String(format: "The minute should be 45, not %d", components.minute!))
        XCTAssertTrue(components.second == 36, String(format: "The seconde should be 36, not %d", components.second!))
        XCTAssertTrue(components.nanosecond == 123001098, String(format: "the nanosecond should be 123001098, not %d", components.nanosecond!))
        XCTAssertTrue((components as NSDateComponents).timeZone! == TimeZone(secondsFromGMT: 3600*7), String(format: "Timezone should be 'GMT+7:00', not %@", (components as NSDateComponents).timeZone! as CVarArg))
      }
    }
  }
  
  func testTimeWithFractionalSeconds() {
    let time = "01:45:36.123"
    
    let parsedDate = Date.parseDateString(time)
    
    XCTAssertNotNil(parsedDate, "Date should not be nil")
    
    if parsedDate != nil {
      var calendar = Calendar.current
      calendar.timeZone = TimeZone(secondsFromGMT: 0)!
      
      let components = (calendar as NSCalendar).components([.hour, .minute, .second, .nanosecond], from: parsedDate!)
      
      XCTAssertTrue(components.hour == 1, String(format: "The hour should be 1, not %d", components.hour!

      ))
      XCTAssertTrue(components.minute == 45, String(format: "The minute should be 45, not %d", components.minute!))
      XCTAssertTrue(components.second == 36, String(format: "The seconde should be 36, not %d", components.second!))
      XCTAssertTrue(components.nanosecond!/100000 == 1230, String(format: "the nanosecond should be 1230, not %d", components.nanosecond!/100000))
    }
  }
  
  func testHoursMinutesAndSeconds() {
    let time = "10:10:13"
    
    let parsedDate = Date.parseDateString(time)
    
    XCTAssertNotNil(parsedDate, "Date should not be nil")
    
    if parsedDate != nil {
      var calendar = Calendar.current
      calendar.timeZone = TimeZone(secondsFromGMT: 0)!
      
      let components = (calendar as NSCalendar).components([.hour, .minute, .second], from: parsedDate!)
      
      XCTAssertTrue(components.hour == 10, String(format: "The hour should be 10, not %d", components.hour!))
      XCTAssertTrue(components.minute == 10, String(format: "The minute should be 10, not %d", components.minute!))
      XCTAssertTrue(components.second == 13, String(format: "The seconde should be 13, not %d", components.second!))
    }
  }
  
  func testHoursMinutes() {
    let time = "17:58"

    if let parsedDate = Date.parseDateString(time) {
      var calendar = Calendar.current
      calendar.timeZone = TimeZone(secondsFromGMT: 0)!
      
      let components = (calendar as NSCalendar).components([.hour, .minute], from: parsedDate)
      
      XCTAssertTrue(components.hour == 17, String(format: "The hour should be 17, not %d", components.hour!))
      XCTAssertTrue(components.minute == 58, String(format: "The minute should be 58, not %d", components.minute!))
    }
  }
  
  func testTimeZoneDesignator() {
    //        Z, +01 or +01:00
    let timeZoneDesignatorZ = "Z"

    if let parsedTimeZoneDesignatorZ = TimeZone.timeZoneWithString(timeZoneDesignatorZ) {
      XCTAssertTrue(parsedTimeZoneDesignatorZ == TimeZone(secondsFromGMT: 0), String(format: "Timezone should be Z, not %@", parsedTimeZoneDesignatorZ as CVarArg))
    }
    
    let timeZonePossitiveWithMinutes = "+05:30" // India

    if let parsedTimeZoneDesignatorWithMinutes = TimeZone.timeZoneWithString(timeZonePossitiveWithMinutes) {
      XCTAssertTrue(parsedTimeZoneDesignatorWithMinutes == TimeZone(secondsFromGMT: ((3600 * 5) + (30 * 60))), String(format: "Timezone should be +5:30, not %@", parsedTimeZoneDesignatorWithMinutes as CVarArg))
    }
    
    let timeZoneNegativeWithoutMinutes = "-07" // India

    if let parsedTimeZoneDesignatorWithoutMinutes = TimeZone.timeZoneWithString(timeZoneNegativeWithoutMinutes) {
      XCTAssertTrue(parsedTimeZoneDesignatorWithoutMinutes == TimeZone(secondsFromGMT: (3600 * -7)), String(format: "Timezone should be -7:00, not %@", parsedTimeZoneDesignatorWithoutMinutes as CVarArg))
    }
  }

  func testDateTime() {
    let dateTime = "2016-10-18 11:58:00"

    if let parsedDateTime = Date.parseDateString(dateTime) {
      var calendar = Calendar.current
      calendar.timeZone = TimeZone(secondsFromGMT: 0)!

      let components = (calendar as NSCalendar).components([.year, .month, .day, .hour, .minute, .second, .nanosecond, .timeZone], from: parsedDateTime)

      XCTAssertTrue(components.year == 2016, String(format: "The year should be 1985, not %d", components.year!))
      XCTAssertTrue(components.month == 10, String(format: "The month should be 4, not %d", components.month!))
      XCTAssertTrue(components.day == 18, String(format: "The day should be 12, not %d", components.day!))
      XCTAssertTrue(components.hour == 11, String(format: "The hour should be 23, not %d", components.hour!))
      XCTAssertTrue(components.minute == 58, String(format: "The minute should be 20, not %d", components.minute!))
      XCTAssertTrue(components.second == 00, String(format: "The seconde should be 50, not %d", components.second!))
    }
  }

  
  func testRFC3339DateTime() {
    let dateTime = "1985-04-12T23:20:50.52Z"

    if let parsedDateTime = Date.parseDateString(dateTime) {
      var calendar = Calendar.current
      calendar.timeZone = TimeZone(secondsFromGMT: 0)!
      
      let components = (calendar as NSCalendar).components([.year, .month, .day, .hour, .minute, .second, .nanosecond, .timeZone], from: parsedDateTime)
      
      XCTAssertTrue(components.year == 1985, String(format: "The year should be 1985, not %d", components.year!))
      XCTAssertTrue(components.month == 04, String(format: "The month should be 4, not %d", components.month!))
      XCTAssertTrue(components.day == 12, String(format: "The day should be 12, not %d", components.day!))
      XCTAssertTrue(components.hour == 23, String(format: "The hour should be 23, not %d", components.hour!))
      XCTAssertTrue(components.minute == 20, String(format: "The minute should be 20, not %d", components.minute!))
      XCTAssertTrue(components.second == 50, String(format: "The seconde should be 50, not %d", components.second!))
      XCTAssertTrue(components.nanosecond == 519999980, String(format: "The nano second should be 519999980, not %d", components.nanosecond!))
      XCTAssertTrue((components as NSDateComponents).timeZone! == TimeZone(secondsFromGMT: 0), String(format: "The timezone should be GMT, not %@", (components as NSDateComponents).timeZone! as CVarArg))
    }
  }
  
  func testRFC2282DateTime() {
    let dateTime = "Fri, 21 Nov 1997 09:55:06 -0600"

    if let parsedDateTime = Date.parseDateString(dateTime) {

      if let timeZone = TimeZone.timeZoneWithString(dateTime) {
        var calendar = Calendar.current
        calendar.timeZone = timeZone

        let components = (calendar as NSCalendar).components([.year, .month, .day, .hour, .minute, .second, .timeZone], from: parsedDateTime)

        XCTAssertTrue(components.year == 1997, String(format: "The year should be 1997, not %d", components.year!))
        XCTAssertTrue(components.month == 11, String(format: "The month should be 11, not %d", components.month!))
        XCTAssertTrue(components.day == 21, String(format: "The day should be 21, not %d", components.day!))
        XCTAssertTrue(components.hour == 9, String(format: "The hour should be 9, not %d", components.hour!))
        XCTAssertTrue(components.minute == 55, String(format: "The minute should be 55, not %d", components.minute!))
        XCTAssertTrue(components.second == 06, String(format: "The seconde should be 06, not %d", components.second!))
        XCTAssertTrue((components as NSDateComponents).timeZone! == TimeZone(secondsFromGMT: (3600 * -6)), String(format: "The timezone should -06.00, not %@", (components as NSDateComponents).timeZone! as CVarArg))
      }
    }
  }
}
